File: apps/ios/WindmillKit/Sources/WindmillGym/Notes.swift
import Foundation
import WindmillPlatform

// A note is a title and a body, stored verbatim, account-only: nothing here is written to the device.
public struct Note: Equatable, Decodable, Sendable, Identifiable {
    public let id: String
    public let position: Int
    public let title: String
    public let body: String
    public let updatedAtMs: Int64

    public init(id: String, position: Int, title: String, body: String, updatedAtMs: Int64 = 0) {
        self.id = id
        self.position = position
        self.title = title
        self.body = body
        self.updatedAtMs = updatedAtMs
    }

    enum CodingKeys: String, CodingKey {
        case id, position, title, body
        case updatedAtMs = "updatedAt"
    }

    public init(from decoder: Decoder) throws {
        let fields = try decoder.container(keyedBy: CodingKeys.self)
        id = try fields.decode(String.self, forKey: .id)
        position = try fields.decodeIfPresent(Int.self, forKey: .position) ?? 0
        title = try fields.decode(String.self, forKey: .title)
        body = try fields.decodeIfPresent(String.self, forKey: .body) ?? ""
        updatedAtMs = try fields.decodeIfPresent(Int64.self, forKey: .updatedAtMs) ?? 0
    }

    // The row's meta: the body's first line, or nothing.
    public var firstLine: String {
        body.split(whereSeparator: \.isNewline).first.map(String.init) ?? ""
    }
}

public struct NoteWrite: Equatable, Encodable, Sendable {
    public let title: String
    public let body: String

    public init(title: String, body: String) {
        self.title = title
        self.body = body
    }
}

// The three bounds are the server's exactly; over-limit is refused here in the server's own sentence.
public enum Notes {
    public static let maxNotes = 10
    public static let maxTitleCharacters = 60
    public static let maxBodyBytes = 500
    // A counter is chrome a short note never carries: each appears in the last fifth of its bound.
    public static let counterFromBytes = 400
    public static let counterFromCharacters = 48

    public static let title = "Notes"
    public static let honesty = "Any agent you connect can read these too."
    public static let purpose = "what you write for Coach"
    public static let precedence = "Top note wins."
    public static let add = "Add a note"
    public static let full = "10 of 10 notes. Delete one to add another."
    public static let placeholders = ["How I want to be talked to", "What I am training for"]

    public static let titleField = "Title"
    public static let bodyField = "What Coach should know"
    public static let save = "Save"
    public static let delete = "Delete note"
    public static let deleteTitle = "Delete this note?"
    public static let deleteConfirm = "Delete"
    public static let keep = "Keep it"

    public static let needsSignIn = "Notes live with your account, so they need you signed in."
    public static let signIn = "Sign in"
    public static let reading = "reading your notes…"
    public static let noAnswer = "the log didn’t answer — nothing changed"

    public static let needsTitle = "a note needs a title"
    public static let titleTooLong = "a title runs to 60 characters"
    public static let bodyTooLong = "a note runs to 500 bytes"

    // The server's alphabet: [A-Za-z0-9_-], the same discipline as a thread id.
    public static func mintNoteId() -> String {
        "note_" + UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
    }

    public static func canAdd(_ count: Int) -> Bool { count < maxNotes }

    // nil while the draft is within bounds; otherwise the sentence the server would answer with.
    public static func refusal(title: String, body: String) -> String? {
        let named = title.trimmingCharacters(in: .whitespacesAndNewlines)
        if named.isEmpty { return needsTitle }
        if titleCharacters(named) > maxTitleCharacters { return titleTooLong }
        if body.trimmingCharacters(in: .whitespacesAndNewlines).utf8.count > maxBodyBytes { return bodyTooLong }
        return nil
    }

    public static func write(title: String, body: String) -> NoteWrite? {
        guard refusal(title: title, body: body) == nil else { return nil }
        return NoteWrite(title: title.trimmingCharacters(in: .whitespacesAndNewlines),
                         body: body.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    public static func bodyBytes(_ body: String) -> Int {
        body.trimmingCharacters(in: .whitespacesAndNewlines).utf8.count
    }

    public static func counter(bytes: Int) -> String? {
        guard bytes >= counterFromBytes else { return nil }
        return "\(bytes) of \(maxBodyBytes) bytes"
    }

    // Code points, not grapheme clusters: `char_length` is what the column checks.
    public static func titleCharacters(_ title: String) -> Int {
        title.trimmingCharacters(in: .whitespacesAndNewlines).unicodeScalars.count
    }

    public static func counter(characters: Int) -> String? {
        guard characters >= counterFromCharacters else { return nil }
        return "\(characters) of \(maxTitleCharacters) characters"
    }
}

// The log's own sentence when it refused, a plain one when it did not answer; never rewritten.
public struct NotesRefusal: Equatable, Error, Sendable {
    public let line: String

    public init(line: String) {
        self.line = line
    }

    public init(_ error: Error) {
        guard let failure = error as? WindmillApiError else {
            self = NotesRefusal(line: Notes.noAnswer)
            return
        }
        switch failure {
        case .offline: self = NotesRefusal(line: failure.line)
        case .malformed: self = NotesRefusal(line: Notes.noAnswer)
        case .refused(_, let refusal): self = NotesRefusal(line: refusal.message ?? Notes.noAnswer)
        }
    }
}

File: apps/android/gym/src/main/kotlin/works/windmill/gym/domain/Notes.kt
package works.windmill.gym.domain

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// /v1/gym/notes, owner-scoped: title-and-body pairs the lifter writes and Coach reads, in precedence
// order. Account-only — this phone keeps no copy and the claim replays none.

@Serializable
data class Note(
    val id: String,
    val position: Int = 0,
    val title: String,
    val body: String = "",
    @SerialName("updatedAt") val updatedAtMs: Long = 0,
) {
    // The row's meta: the first line of the body that says anything.
    val firstLine: String?
        get() = body.lineSequence().map { it.trim() }.firstOrNull { it.isNotEmpty() }
}

// Both fields carry NO DEFAULT: encodeDefaults is off, so an empty body would travel as an absent key
// and the route reads an absent key as a note it cannot read.
@Serializable
data class NoteWrite(val title: String, val body: String)

@Serializable
data class NotesOrder(val order: List<String>)

object Notes {
    const val title = "Notes"

    // The same three numbers the server checks; the server's sentence is what a refusal shows.
    const val maxNotes = 10
    const val titleMax = 60
    const val bodyMaxBytes = 500
    // A counter is chrome only in the last fifth of its bound: a short note carries none.
    const val counterFrom = 400
    const val titleCounterFrom = 48

    // The head, and nothing else in it.
    const val honesty = "Any agent you connect can read these too."
    const val sub = "what you write for Coach"

    // The one caption on the screen: the handle shows the drag, this says what the order means.
    const val topWins = "Top note wins."

    const val add = "Add a note"
    const val full = "10 of 10 notes. Delete one to add another."

    // Placeholder text inside empty rows, never stored: tapping one opens the editor with the title
    // filled in, and nothing is written until the lifter saves.
    val placeholders = listOf("How I want to be talked to", "What I am training for")

    const val signedOut = "Notes live with your account, so they need you signed in."

    // Beside the dials in settings, where the question is asked.
    const val settingsLine = "Coach reads your notes, not your settings."

    const val titlePlaceholder = "Title"
    const val bodyPlaceholder = "What Coach should know"
    const val save = "Save"
    // One tap. The note leaves the editor and the window holds it, with Undo on the room's transient.
    const val delete = "Delete note"

    fun bytes(body: String): Int = body.toByteArray(Charsets.UTF_8).size

    // Null below the threshold: nothing is drawn.
    fun counter(body: String): String? {
        val used = bytes(body.trim())
        if (used < counterFrom) return null
        return "$used of $bodyMaxBytes bytes"
    }

    // Past the bound the counter goes alarm; Save stays tappable and the log's sentence refuses in place.
    fun over(body: String): Boolean = bytes(body.trim()) > bodyMaxBytes

    // Code points, not UTF-16 units: `char_length` is what the column checks.
    fun titleLength(title: String): Int = title.trim().let { it.codePointCount(0, it.length) }

    fun titleCounter(title: String): String? {
        val used = titleLength(title)
        if (used < titleCounterFrom) return null
        return "$used of $titleMax characters"
    }

    fun titleOver(title: String): Boolean = titleLength(title) > titleMax

    fun savable(title: String): Boolean = title.trim().isNotEmpty()

    // One step at a time, the way the drag hands it over.
    fun moved(notes: List<Note>, from: Int, to: Int): List<Note> {
        if (from == to || from !in notes.indices || to !in notes.indices) return notes
        val order = notes.toMutableList()
        order.add(to, order.removeAt(from))
        return order
    }
}
